dask/array/core.py

```diff
@@ -927,7 +928,11 @@ def finalize(results):
     results2 = results
     while isinstance(results2, (tuple, list)):
         if len(results2) > 1:
-            return concatenate3(results)
+            x = unpack_singleton(results)
+            if inspect.getmodule(x):
+                return _concatenate2(results, axes=list(range(x.ndim)))
+            else:
+                return concatenate3(results)
```

Comment:
Can you explain this branch? Looks like its for numpy vs sparse (a comment would help)? Why is `concatenate3` not supported for sparse?